### PR TITLE
[release/v2.13] Add configurable default addon enforce option

### LINKED ISSUE #4979: enforcement of default addons in user cluster
**User Story**
As Kubermatic operator I want to enforce that some addons are placed in the cluster. Even if the user cluster admin deletes a addon it should get recreated.

Currently Kubermatic only recreates the addons when a cluster definition is changed. In my opinion the addon controller should continuously check if there is a diff between addon config and addon deployment
This behavior was tested at the  `weekly-2020-03` version. 

**Acceptance criteria**
- [ ] default addons get enforced
- [ ] a deleted default addon in the user cluster get created immediately

### PR DESCRIPTION
**Pull request by moelsayed**
**What this PR does / why we need it**:
Cherry-pick configurable default addon enforce option, along with related code
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4979

**Special notes for your reviewer**:
This PR extracts the basic logic from #5193, but since v2.13.x doesn't have a generated list of addons, the PR only adds the enforce labels only when addons are added using a list. When an addons file is used, the user should add the labels manually to the yaml Addon file.

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Add a configuration flag for seed-controller-manager to enforce default addons on userclusters. Disabled by default.
```

### CODE REVIEW

File: .prow.yaml [hunk 1]
```diff
@@ -292,7 +292,7 @@ presubmits:
   # e2e tests for Kubernetes 1.15
   #########################################################
 
-  - name: pre-kubermatic-e2e-aws-coreos-1.15
```
Comment (irozzo-1A) [inline .prow.yaml:295]: We switched to `flatcar` in 2.14 and 2.15, maybe we should do the same for consistency.

File: api/pkg/resources/address/address_test.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ const (
 	fakeClusterNamespaceName = "cluster-ns"
 	// This must be the A record for *.europe-west3c.dev.kubermatic.io and
 	// *.alias-europe-west3-c.dev.kubermatic.io
-	externalIP = "35.198.93.90"
```
Comment (irozzo-1A) [inline api/pkg/resources/address/address_test.go:25]: I'm cherry picking the proper fix for this here: https://github.com/kubermatic/kubermatic/pull/6020

File: .prow.yaml (hunk 1 above)
Comment (irozzo-1A) [reply .prow.yaml:295]: well, probably in v2.13 was not supported yet

Comment (irozzo-1A) [reply .prow.yaml:295]: Yep, I confirm.

File: api/pkg/resources/address/address_test.go (hunk 2 above)
Comment (moelsayed) [reply api/pkg/resources/address/address_test.go:25]: Sounds good. I am going to keep it though until that PR is merged, otherwise I won't pass tests.

Comment (irozzo-1A) [reply api/pkg/resources/address/address_test.go:25]: sure

**Comment by moelsayed** [comment]
Since this is considered a change in behavior, not technically a bug fix. I am updating the charts to have it disabled by default so avoid the change in behavior for customers who don't need this.

File: api/pkg/controller/seed-controller-manager/addon/addon_controller.go [hunk 3]
```diff
@@ -242,7 +259,9 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, addo
 	if err := r.ensureFinalizerIsSet(ctx, addon); err != nil {
 		return nil, fmt.Errorf("failed to ensure that the cleanup finalizer existis on the addon: %v", err)
 	}
-
+	if err := r.ensureResourcesCreatedConditionIsSet(ctx, addon); err != nil {
+		return nil, fmt.Errorf("failed to set add ResourcesCreated Condition: %v", err)
```
Comment (xrstf) [inline api/pkg/controller/seed-controller-manager/addon/addon_controller.go]: `set add` should be either `set` or `add` :)

File: api/pkg/controller/seed-controller-manager/addon/addon_controller.go [hunk 4]
```diff
@@ -540,3 +559,53 @@ func (r *Reconciler) ensureRequiredResourceTypesExist(ctx context.Context, log *
...
+	}
+	return -1, nil
+}
+
+func addonResourcesCreated(addon *kubermaticv1.Addon) bool {
+	_, cond := getAddonCondition(addon, kubermaticv1.AddonResourcesCreated)
+	if cond != nil && cond.Status == corev1.ConditionTrue {
+		return true
+	}
+	return false
```
Comment (xrstf) [inline api/pkg/controller/seed-controller-manager/addon/addon_controller.go]: This can be written as a single `return` statement.

**Review by xrstf — APPROVED** [review]
/approve
/hold

Two minor things that could be improved, but since this is just a backport, I'm fine with the current state also. :-) Please unhold when @irozzo-1A if @irozzo-1A is also good with this.

**Pull request merged**